File: java_bench/src/Parser.java
import java.util.*;


public class Parser {
    public int limit;

    public Parser(int limit) {
        this.limit = limit;
    }

    Result parse(String text) {
        Counter<String> wordCounter = new Counter<>();
        Counter<Character> letterCounter = new Counter<>();
        for (String word : text.toLowerCase().split("[\\s,.?!]+")) {
            wordCounter.add(word);
            for (Character letter : word.toCharArray()) {
                letterCounter.add(letter);
            }
        }
        return new Result(wordCounter.top(limit), letterCounter.top(limit));
    }


    public record Result (
        Map<String, Integer> topWords,
        Map<Character, Integer> topLetters
    ) {}


    static class Counter<T> {
        Map<T, Integer> _map = new HashMap<>();

        void add(T item) {
            int count = _map.getOrDefault(item, 0);
            _map.put(item, count + 1);
        }

        Map<T, Integer> top(int limit) {
            Map<T, Integer> result = new LinkedHashMap<>();
            _map.entrySet().stream()
                    .sorted((a, b) -> b.getValue().compareTo(a.getValue()))
                    .limit(limit)
                    .forEach(e -> result.put(e.getKey(), e.getValue()));
            return result;
        }
    }
}

File: kotlin_bench/src/main/kotlin/thinkwithme/Parser.kt
package thinkwithme

import java.util.*

class Parser(var limit: Int) {
    fun parse(text: String): Result {
        val wordCounter = Counter<String>()
        val letterCounter = Counter<Char>()
        for (word in text.lowercase(Locale.getDefault()).split("[\\s,.?!]+".toRegex()).dropLastWhile { it.isEmpty() }
            .toTypedArray()) {
            wordCounter.add(word)
            for (letter in word.toCharArray()) {
                letterCounter.add(letter)
            }
        }
        return Result(wordCounter.top(limit), letterCounter.top(limit))
    }


    data class Result(val topWords: Map<String, Int>, val topLetters: Map<Char, Int>)


    internal class Counter<T> {
        private val _map: MutableMap<T, Int> = HashMap()

        fun add(item: T) {
            val count = _map.getOrDefault(item, 0)
            _map[item] = count + 1
        }

        fun top(limit: Int): Map<T, Int> {
            val result: MutableMap<T, Int> = LinkedHashMap()
            _map.entries.stream()
                .sorted { a: Map.Entry<T, Int>, b: Map.Entry<T, Int> -> b.value.compareTo(a.value) }
                .limit(limit.toLong())
                .forEach { e: Map.Entry<T, Int> -> result[e.key] = e.value }
            return result
        }
    }
}